### PR TITLE
active storage validation classes added

### PR DESCRIPTION
**Pull request by reckerswartz**
### Summary
Alert not completed guidance required
PR with reference #33741 
 ```activestorage/lib/active_storage/validations.rb``` not wriiten to called in ```activestorage/lib/active_storage.rb```
## Usage

For example you have a model like this and you want to add validation.

```ruby
class User < ApplicationRecord
  has_one_attached :avatar
  has_many_attached :photos

  validates :name, presence: true

  validates :avatar, attached: true, content_type: 'image/png'
  validates :photos, attached: true, content_type: ['image/png', 'image/jpg', 'image/jpeg']
end
```

or

```ruby
class Project < ApplicationRecord
  has_one_attached :preview
  has_one_attached :attachment

  validates :title, presence: true

  validates :preview, attached: true, size: { less_than: 100.megabytes , message: 'is not given between size' }
  validates :attachment, attached: true, content_type: { in: 'application/pdf', message: 'is not a PDF' }
end
```

## Internationalization (I18n)

Active Storage Validations use I18n for errors messages. For this add there keys in your translation file :

```yml
en:
  errors:
    messages:
      content_type_invalid: "has an invalid content type"
```

### CODE REVIEW

**Comment by reckerswartz** [comment]
r?

File: activestorage/lib/active_storage/validations/size_validator.rb [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+module ActiveStorage
+  module Validations
+    class SizeValidator < ActiveModel::EachValidator
+      delegate :number_to_human_size, to: ActiveSupport::NumberHelper
+
+      AVAILABLE_CHECKS = [:less_than, :less_than_or_equal_to, :greater_than, :greater_than_or_equal_to, :between]
+
+      def check_validity!
+        unless (AVAILABLE_CHECKS).any? { |argument| options.has_key?(argument) }
+          raise ArgumentError, "You must pass either :less_than, :greater_than, or :between to the validator"
```
Comment (connorshea) [inline activestorage/lib/active_storage/validations/size_validator.rb]: Should this include all of the `AVAILABLE_CHECKS`? Or more realistically just use these as examples and keep the error message as "You must pass a validation check to the validator."

File: activestorage/lib/active_storage/validations/size_validator.rb [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+        files.each do |file|
+          if content_size_valid?(file)
+            record.errors.add(attribute, options[:message].presence || "size #{number_to_human_size(file.blob.byte_size)} is not between required range")
+            return
+          end
+        end
+      end
+
+      def content_size_valid?(file)
```
Comment (connorshea) [inline activestorage/lib/active_storage/validations/size_validator.rb]: It's not immediately obvious to me what `size: { greater_than: 100.kilobytes }` does, I assume it means the minimum file size is 100kb?

I find the DSL here kind of confusing, compare the ActiveRecord validations on length with these ActiveStorage file size validations:

```ruby
# ActiveRecord
validates :name, length: { minimum: 4, maximum: 20 }
validates :name, length: { minimum: 2 }
validates :bio, length: { maximum: 500 }
validates :password, length: { in: 6..20 }
validates :registration_number, length: { is: 6 }

# ActiveStorage
validates :preview, attached: true, size: { greater_than: 100.kilobytes }
validates :preview, attached: true, size: { between: 0..100.kilobytes }
validates :preview, attached: true, size: { less_than: 100.megabytes }
```

Shouldn't the ActiveStorage validators use `maximum`, `minimum`, and `in` like other validations? (I don't think `is` would be very useful for file size).

Comment (connorshea) [reply activestorage/lib/active_storage/validations/size_validator.rb]: I guess I don’t really see the usefulness of >= or <=. As for between, wouldn’t that be covered by `in`?

**Comment by igorkasyanchuk** [comment]
I've one comment that soon I'll merge this PR with one more validation, thanks to @ivanelrey 
https://github.com/igorkasyanchuk/active_storage_validations/pull/16
it's also could be ported to rails 6

**Comment by abhchand** [comment]
@reckerswartz / @igorkasyanchuk / all - 

Great work on this PR so far! I actually went down the road of implementing this myself before I realized I had missed this PR and it was already in progress. 

I think there's a few things missing from this PR. I'm not a maintainer on this project, but if I were I wouldn't merge it for a few reasons (outlined below)

It's been over a month here so I'm not sure if this is still active. If you'd like, I can open a PR very soon or even build off this one a bit.

Some notes - 

1. Tests - I didn't see any added here at all. This can (and should!) be tested for scenarios like valid/invalid, one attachment, many attachments, etc...

2. Namespacing - The validations should be namespaced with `attachment_` or similar. Otherwise we're claiming a name like `SizeValidator` that no other parts of the library or external gems can use. 

3. Shortcuts / Helpers - Taking a cue from [thoughtbot's paperclip library](https://github.com/thoughtbot/paperclip#validations) (which is now deprecated in favor of activestorage): this could easily support helpers like `validate_attachment_presence :avatar` and `validate_attachment :avatar, presence: true`. These are straightforward to add and conform to the pattern we see in other Rails libraries. 

4. Documentation - I don't see any additions to the changelog or mention of validations in the README or Rails guides. Formatting of the commit messages also does not provide much detail or conform to Rails standards.

Despite the above, great work so far on adding this feature 👍 . Just my two cents!

👋

**Comment by igorkasyanchuk** [comment]
@abhchand thank you.
Maybe you can join @reckerswartz and do it together. Now I'm a bit busy with other activities. 
I guess @reckerswartz could easily add you as a contributor to his fork if you like.

File: activestorage/lib/active_storage/validations/attached_validator.rb [hunk 3]
```diff
@@ -0,0 +1,13 @@
+# frozen_string_literal: true
+
+module ActiveStorage
+  module Validations
+    class AttachedValidator < ActiveModel::EachValidator
+      def validate_each(record, attribute, value)
+        unless record.send(attribute).attached?
```
Comment (abhchand) [inline activestorage/lib/active_storage/validations/attached_validator.rb]: `attached?` just looks to see if an associated `ActiveStorage::Attachment` record is present ([see here](https://github.com/rails/rails/blob/master/activestorage/lib/active_storage/attached/one.rb#L46))

On creation, an attachment is only persisted/saved *after* the model itself has been persisted/saved ([see here](https://github.com/rails/rails/blob/master/activestorage/lib/active_storage/attached/model.rb#L54)). Until then, it's simply noted in `attachment_changes`. 

So basically this validation will always fail on create because at that time a record will never have an associated blob saved. 

One (potential) way around this is too perform a different check if the record is `persisted?` or not. If not persisted then check that `attachment_changes1 at least contains a promise that an attachment *will* be saved. 

Would love if someone with more in depth knowledge could confirm the above.

File: activestorage/lib/active_storage/validations/size_validator.rb [hunk 4]
```diff
@@ -0,0 +1,48 @@
+# frozen_string_literal: true
+
+module ActiveStorage
+  module Validations
+    class SizeValidator < ActiveModel::EachValidator
+      delegate :number_to_human_size, to: ActiveSupport::NumberHelper
+
+      AVAILABLE_CHECKS = [:less_than, :less_than_or_equal_to, :greater_than, :greater_than_or_equal_to, :between]
```
Comment (abhchand) [inline activestorage/lib/active_storage/validations/size_validator.rb]: From @connorshea's comments below, this could be greatly simplified to 

- `:minimum`
- `:maximum`
- `:in`

File: activestorage/lib/active_storage/validations/attached_validator.rb (hunk 3 above)
Comment (abhchand) [reply activestorage/lib/active_storage/validations/attached_validator.rb]: Also if true, scenarios like the above are why we need tests around all major combinations of these features 😃

**Review by abhchand — CHANGES_REQUESTED** [review]

File: activestorage/lib/active_storage/validations/attached_validator.rb (hunk 3 above)
Comment (JoeWoodward) [reply activestorage/lib/active_storage/validations/attached_validator.rb]: Should also use a guard statement here. 
```ruby
return if record.send(attribute).attached?
record.errors.add(attribute, :blank)
```

File: activestorage/lib/active_storage/validations/content_type_validator.rb [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+        return true unless files.attached?
+        return true if types.empty?
+
+        files = Array.wrap(files)
+
+        errors_options = { authorized_types: types_to_human_format }
+        errors_options[:message] = options[:message] if options[:message].present?
+
+        files.each do |file|
+          unless content_type_valid?(file)
```
Comment (JoeWoodward) [inline activestorage/lib/active_storage/validations/content_type_validator.rb]: Remove unnecessary indentation with a guard clause
```ruby
next if content_type_valid?(file)
...
```

File: activestorage/lib/active_storage/validations/content_type_validator.rb [hunk 6]
```diff
@@ -0,0 +1,44 @@
+# frozen_string_literal: true
+
+module ActiveStorage
+  module Validations
+    class ContentTypeValidator < ActiveModel::EachValidator
+      def validate_each(record, attribute, value)
+        files = record.send(attribute)
+
+        return true unless files.attached?
```
Comment (JoeWoodward) [inline activestorage/lib/active_storage/validations/content_type_validator.rb]:
```ruby
return true unless files.attached? && types.present?
```

File: activestorage/lib/active_storage/validations/size_validator.rb [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+
+      def validate_each(record, attribute, value)
+        files = record.send(attribute)
+        # only attached
+        return true unless files.attached?
+
+        files = Array.wrap(files)
+
+        files.each do |file|
+          if content_size_valid?(file)
```
Comment (JoeWoodward) [inline activestorage/lib/active_storage/validations/size_validator.rb]:
```ruby
next unless content_size_valid?(file)
...
```

File: activestorage/lib/active_storage/validations/size_validator.rb (hunk 2 above)
Comment (JoeWoodward) [reply activestorage/lib/active_storage/validations/size_validator.rb]: I agree. When comparing numericality it makes sense, in mathematics we need these functions, however, when validating a file size it's a little overkill. IRL I would never specify `this file more than 100kb` (101kb or more). It's always going to be `should be at least 100kb no more than 1mb`.

Comment (JoeWoodward) [reply activestorage/lib/active_storage/validations/size_validator.rb]: I think it's actually not useful to have less/greater than. maximum/minimum are replacements for less/greater_than_or_equal_to

e.g.
`file max 100kb` is the same as `file <= 100kb` not `file < 100kb`

File: activestorage/lib/active_storage/validations/content_type_validator.rb [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+
+        errors_options = { authorized_types: types_to_human_format }
+        errors_options[:message] = options[:message] if options[:message].present?
+
+        files.each do |file|
+          unless content_type_valid?(file)
+            errors_options[:content_type] = content_type(file)
+
+            record.errors.add(attribute, :content_type_invalid, errors_options)
+            return
```
Comment (JoeWoodward) [inline activestorage/lib/active_storage/validations/content_type_validator.rb]: Is this supposed to return here? Is this so that the error is only added once per input?

Comment (reckerswartz) [reply activestorage/lib/active_storage/validations/content_type_validator.rb]: Hey sorry! code is updated in Gem. I will update here soon or open new PR.
https://github.com/igorkasyanchuk/active_storage_validations/tree/master/lib/active_storage_validations

**Comment by reckerswartz** [comment]
@JoeWoodward many thanks for your efforts but I think it's time to open new PR with many things updated
I will ask you suggest things over there.

File: activestorage/lib/active_storage/validations/attached_validator.rb (hunk 3 above)
Comment (JoeWoodward) [reply activestorage/lib/active_storage/validations/attached_validator.rb]: why not use `value.attached?`?

**Comment by abhchand** [comment]
@JoeWoodward - I opened a similar PR https://github.com/rails/rails/pull/35390 now that this is closed. Appreciate your review here - any chance you could take a look at that as well?
